provider/redis.go

```diff
@@ -34,58 +38,95 @@ func redisOnlineStoreFactory(serialized pc.SerializedConfig) (Provider, error) {
+func NewRedisOnlineStore(options *pc.RedisConfig) (*redisOnlineStore, error) {
+	redisOptions := rueidis.ClientOption{
+		InitAddress: []string{options.Addr},
+		Password:    options.Password,
+		SelectDB:    options.DB,
+	}
+	redisClient, err := rueidis.NewClient(redisOptions)
+	if err != nil {
+		return nil, err
 	}
-	redisClient := redis.NewClient(redisOptions)
 	return &redisOnlineStore{redisClient, options.Prefix, BaseProvider{
 		ProviderType:   pt.RedisOnline,
 		ProviderConfig: options.Serialized(),
 	},
-	}
+	}, nil
 }
 
 func (store *redisOnlineStore) AsOnlineStore() (OnlineStore, error) {
 	return store, nil
 }
 
+func (store *redisOnlineStore) AsVectorStore() (VectorStore, error) {
+	return store, nil
+}
+
 func (store *redisOnlineStore) Close() error {
-	// from the docs:
-	//
-	// "It is rare to Close a Client, as the Client is meant to be
-	// long-lived and shared between many goroutines.""
-	//
-	return store.client.Close()
+	store.client.Close()
+	return nil
 }
 
 func (store *redisOnlineStore) GetTable(feature, variant string) (OnlineStoreTable, error) {
 	key := redisTableKey{store.prefix, feature, variant}
-	vType, err := store.client.HGet(context.TODO(), fmt.Sprintf("%s__tables", store.prefix), key.String()).Result()
+	cmd := store.client.B().
+		Hget().
+		Key(fmt.Sprintf("%s__tables", store.prefix)).
+		Field(key.String()).
+		Build()
+	vType, err := store.client.Do(context.TODO(), cmd).ToString()
 	if err != nil {
 		return nil, &TableNotFound{feature, variant}
 	}
-	table := &redisOnlineTable{client: store.client, key: key, valueType: ValueType(vType)}
+	var valueType ValueType
+	if _, isScalarType := ScalarTypes[ScalarType(vType)]; isScalarType {
+		valueType = ScalarType(vType)
+	} else {
+		valueType = &VectorType{}
+		err = json.Unmarshal([]byte(vType), valueType)
+		if err != nil {
+			return nil, err
+		}
+	}
+	table := &redisOnlineIndex{
+		client:    store.client,
+		key:       key,
+		valueType: valueType,
+	}
 	return table, nil
 }
 
 func (store *redisOnlineStore) CreateTable(feature, variant string, valueType ValueType) (OnlineStoreTable, error) {
 	key := redisTableKey{store.prefix, feature, variant}
-	exists, err := store.client.HExists(context.TODO(), fmt.Sprintf("%s__tables", store.prefix), key.String()).Result()
+	cmd := store.client.B().
+		Hexists().
+		Key(fmt.Sprintf("%s__tables", store.prefix)).
+		Field(key.String()).
+		Build()
+	exists, err := store.client.Do(context.TODO(), cmd).AsBool()
 	if err != nil {
 		return nil, err
 	}
 	if exists {
 		return nil, &TableAlreadyExists{feature, variant}
 	}
-	if err := store.client.HSet(context.TODO(), fmt.Sprintf("%s__tables", store.prefix), key.String(), string(valueType)).Err(); err != nil {
+	serialized, err := json.Marshal(ValueTypeJSONWrapper{valueType})
+	if err != nil {
 		return nil, err
 	}
+	cmd = store.client.B().
+		Hset().
+		Key(fmt.Sprintf("%s__tables", store.prefix)).
+		FieldValue().
+		FieldValue(key.String(), string(serialized)).
```

Comment:
Is this backwards compatible? What happens if something was in a cluster where it was stored the other way?

Reply:
So, if you look at how I'm handling the value type logic in `GetTable`, I believe it's backwards compatible. Let's say a user has a preexisting Redis instance with the value `int32`. I'll first try to determine if this is a "legacy" scalar value, and if it is, I'll use that. Otherwise, I'll try to unmarshal the string to `ValueType` and then determine the instance type of that. Full disclosure, this explanation isn't actually what you see above as I was pushing to test this last night and didn't actually implement completely. I'm add this now.

Reply:
This is now working and covered by unit tests.